doc/man/man3/crypto_kex_xk1_init_client.3monocypher

```diff
@@ -0,0 +1,274 @@
+.It Fa remote_pk
+The public key of the server, which must be known to the client before
+making a connection.
+This is used to authenticate the server's identity.
+.El
+.Pp
+After creating initializing
+.Fa ctx , 
+the client calls
+.Fn crypto_kex_xk1_1
+to generate
+.Fa msg1 .
+It proceeds to send
+.Fa msg1
+to the server.
+The client then reads
+.Fa msg2
+from the server.
+It calls
+.Fn crypto_kex_xk1_3 ,
+which yields the
+.Fa session_key
+and
+.Fa msg3 .
+.Fn crypto_kex_xk1_3
+returns zero on success and -1 if the server's message mismatches the
+key exchange until now.
+.Em Always check the return value .
+On success, this function also wipes the data in
+.Fa ctx .
+The client sends
+.Fa msg3
+to the server to complete the key exchange.
+.Ss Key exchange as the server
+When accepting a new connection from a client, the server first calls
+.Fn crypto_client_xk1_init_server
+to initialize a
+.Vt crypto_kex_ctx .
+The arguments to
+.Fn crypto_client_xk1_init_server
+are identical to the arguments to
+.Fn crypto_client_xk1_init_client ,
+except that the server does not know a client's public key yet,
+so no
+.Fa remote_pk
+can be specified;
+the server's
+.Fa local_sk
+and
+.Fa local_pk
+are generated ahead of time and known to the clients.
+.Pp
+After initializing
+.Fa ctx , 
+the server waits for the client to send
+.Fa msg1
+and reads it from the network.
+The server then calls
+.Fn crypto_kex_xk1_2
+with
+.Fa msg1 ,
+which yields
+.Fa msg2 .
+.Fa msg2
+is sent to the client.
+The server reads the client's response
+.Fa ( msg3 )
+from the network and calls
+.Fn crypto_kex_xk1_4
+with
+.Fa msg3 ,
+which yields
+.Fa session_key
+and
+.Fa remote_pk ,
+the client's public key.
+.Fn crypto_kex_xk1_4
+returns zero on success and -1 if the client's message mismatches the
+key exchange until now.
+.Em Always check the return value .
+On success, this function also wipes the data in
+.Fa ctx .
+.Sh RETURN VALUES
+.Fn crypto_kex_xk1_init_client ,
+.Fn crypto_kex_xk1_init_server ,
+.Fn crypto_kex_xk1_1
+and
+.Fn crypto_kex_xk1_2
+return nothing.
+.Pp
+.Fn crypto_kex_xk1_3
+and
+.Fn crypto_kex_xk1_4
+return 0 on success or -1 if the messages were corrupt or mismatched the
+server's identity.
+Corruption can be caused by transmission errors, programmer error, or an
+attacker's interference.
+.Sh EXAMPLES
+These examples use UNIX\(rg networking functions to keep the examples
+readable,
+but any kind of networking code will work.
+It is assumed that the client and server have already established a
+connection with each other.
+.Pp
+Client:
+.Bd -literal -offset indent
+crypto_kex_ctx client_ctx; /* key exchange context */
+uint8_t sk[32];            /* 32 random bytes ephemeral secret key */
+uint8_t their_pk[32];      /* the server's public key 
+                            * (known ahead of time) */
+uint8_t seed[32];          /* 32 random bytes */
+uint8_t buf[48];           /* buffer for the key exchange messages */
```

Comment:
Okay, we have *just* established that the kex is a large enough footgun that I butcher the first attempt at documenting them (the X pattern a bit more than the XK1 pattern), and you're proposing to leave a minute detail that can yield an amplification attack up to the application's discretion. Let's not go there and go with a constant message size of 48 bytes. We should give as many correct right answers as we can when the tradeoff is reasonable.

With all due respect to people writing tiny embedded systems and their constraints: 16 additional bytes simply must be available, especially since they're needed later on in the kex anyway.

Reply:
OK, I'll modify the code, and zero-pad the first request. I'll also update the Monokex generator accordingly. Now we have another problem, though: people might be tempted to put *data* in there. I don't want the server to check the contents of this padding, because it would mean a return code where none is needed, but well…

I guess if we act as if all 48 bytes are needed in the API and the documentation, naive users will not have any funny idea?

Reply:
Just in case, add a warning in the comment that this was done deliberately.